compiler-core/src/type_/expression.rs

```diff
@@ -1046,6 +1046,39 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
                 }
             }
 
+            ClauseGuard::FieldAccess {
+                location: _,
+                label,
+                container,
+                ..
+            } => {
+                if let ClauseGuard::Var { location, name, .. } = *container.clone() {
+                    let var_expr = UntypedExpr::Var { location, name };
+                    match self.infer_field_access(
+                        var_expr,
+                        label.clone(),
+                        location,
+                        FieldAccessUsage::Other,
+                    )? {
+                        TypedExpr::RecordAccess { index, typ, .. } => {
+                            let var = self.infer_clause_guard(*container)?;
+                            Ok(ClauseGuard::FieldAccess {
+                                location,
+                                label,
+                                index: Some(index + 1),
+                                type_: typ,
+                                container: Box::new(var),
+                            })
+                        }
+
+                        _ => panic!("Expected RecordAccess"),
```

Comment:
We shouldn't make assertions about the variant returned here, rather any shared code should be agnostic to which AST is being used. We can largely adapt `infer_known_record_access` to be agnostic and have a small wrapper for each kind of AST.